Realm,FSCalendar実装

oggysy opened the pull request:
以下機能を追加しました。
・Realmにデータを保存
・FSCalendar画面を作成
流すアクションをすると、該当日のカレンダーに流した時に表示されていたイラスト画像を表示させる
(一日に複数回流すアクションをしても最初にその日の最初に流すアクションをした時の画像が表示される)
・ロープの横にスワイプを促すLottiAnimationを追加
・Viewの更新をTimerを使って10分ごとに実行
・ロード画面の作成
・アプリアイコンの追加

oggysy committed:
Realm,FSCalendar実装

inushima commented on unoji/Model/WashTimeModel.swift:
```diff
@@ -6,36 +6,52 @@
-        return Int(abs(washTime.timeIntervalSinceNow))
+class WashTimeModel: Object {
+    @objc dynamic var washTime: Date? // 流すアクションをした時刻を記録
+    @objc dynamic var washDay: String = ""// FSCalender用に年月日を文字列に
```
宣言時にデフォルト値を指定した変数は型推論が効くので、明示的な型宣言はつけないことが多いです。

```suggestion
    @objc dynamic var washDay: String = ""// FSCalender用に年月日を文字列に
 
Add a suggestion, <Cmd+g>
Add heading textAdd bold text, <Cmd+b>Add italic text, <Cmd+i>
Add a quote, <Cmd+Shift+.>Add code, <Cmd+e>Add a link, <Cmd+k>
Add a bulleted list, <Cmd+Shift+8>Add a numbered list, <Cmd+Shift+7>Add a task list, <Cmd+Shift+l>
Directly mention a user or team
Reference an issue, pull request, or discussion
Add saved reply
Slash commands
宣言時にデフォルト値を指定した変数は型推論が効くので、明示的な型宣言はつけないことが多いです。

```swift
    @objc dynamic var washDay = ""// FSCalender用に年月日を文字列に
    @objc dynamic var elapsedTime = 0// 経過時間
    @objc dynamic var displayIllustration = "baby"// 経過時間に応じたイラスト名を文字列で
```

inushima replied:
washDayはwashTimeから算出できるので属性とせず、washTimeからwashDayをフォーマットするメソッドを定義するのがよさそうです。

ある属性から算出可能な属性をModelの項目にすると2つの属性整合性を保たないとバグになり、ロジックが複雑化するためです。

inushima commented on unoji/Model/WashTimeModel.swift:
```diff
@@ -6,36 +6,52 @@
+class WashTimeModel: Object {
+    @objc dynamic var washTime: Date? // 流すアクションをした時刻を記録
+    @objc dynamic var washDay: String = ""// FSCalender用に年月日を文字列に
+    @objc dynamic var elapsedTime: Int = 0// 経過時間
```
宣言時にデフォルト値を指定した変数は型推論が効くので、明示的な型宣言はつけないことが多いです。

```suggestion
    @objc dynamic var washDay = ""// FSCalender用に年月日を文字列に
    @objc dynamic var elapsedTime = 0// 経過時間
    @objc dynamic var displayIllustration = "baby"// 経過時間に応じたイラスト名を文字列で
```

inushima commented on unoji/Model/WashTimeModel.swift:
```diff
@@ -6,36 +6,52 @@
+
+extension WashTimeModel {
+    // 最後にRealmに記録されたデータを元にelapsedTimeとdisplayIllustrationの値を更新する
+    func setValue(lastTime: Date) {
```
[nits]
メソッド名はコメントを見なくても中身が類推できるように命名することが一般的です。
Swiftの場合は、メソッド名と引数名が使えるので、以下のようにすると
lastTimeによって関連する属性を更新メソッドであることが分かります。

```suggestion
    func update(by lastTime: Date) {
```

inushima commented on unoji/Model/WashTimeModel.swift:
```diff
@@ -6,36 +6,52 @@
+    @objc dynamic var washTime: Date? // 流すアクションをした時刻を記録
+    @objc dynamic var washDay: String = ""// FSCalender用に年月日を文字列に
+    @objc dynamic var elapsedTime: Int = 0// 経過時間
+    @objc dynamic var displayIllustration: String = "baby"// 経過時間に応じたイラスト名を文字列で
```
displayIllustrationもelapsedTimeから算出できるので、Modelの属性として保存せず、elapsedTimeから都度算出するメソッドとして定義するのがよさそうです。

inushima commented on unoji/Controller/ViewController.swift:
```diff
@@ -93,12 +107,40 @@ class ViewController: UIViewController {
+        //　この関数が呼ばれたらtimerをストップ
+        timer.invalidate()
+        // 変数lastTimeに前回のwashTime変数の値を代入する
+        guard let lastTime = realm.objects(WashTimeModel.self).last?.washTime else { //　nilの場合は現在の時間を入れてsetValue関数を呼び、Viewを更新
```
前回の記録がない場合とある場合の違いは、lastTimeがDate()とするかrealmの保存したデータかの違いで、他のロジックは同じに見えます。
lastTimeの定義を以下のようにすると、gaurdとelse文の中身が不要になり、ロジックが共通化できると思います。

```swift
        let lastTime = realm.objects(WashTimeModel.self).last?.washTime ?? Date()
```

inushima commented on unoji/Controller/ViewController.swift:
```diff
@@ -51,8 +60,12 @@ class ViewController: UIViewController {
-                // アクションの時間をModelに記録
+                //
+                updateView()
                 washTimeModel.washTime = Date()
```
66行目のsetWashTime()でwashTimeの更新をしているので、65行目は不要かと思います。
また、Modelの中身の更新ロジックはできる限りModelに閉じるべきなので、Modelの属性をViewから直接更新することは避けるほうがよいですね。

inushima commented on unoji/Controller/CalendarViewController.swift:
```diff
@@ -0,0 +1,65 @@
+        configure()
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
```
モバイルアプリでデータ更新を画面に反映させる基本的な考え方ですが、画面のリロードは極力行わず、データの更新を監視して画面に反映します。


```
import UIKit
import FSCalendar
import RealmSwift

class ViewController: UIViewController, FSCalendarDelegate, FSCalendarDataSource {
    
    @IBOutlet weak var calendar: FSCalendar!
    private var realm: Realm!
    private var events: Results<MyEvent>!
    private var notificationToken: NotificationToken?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        realm = try! Realm()
        events = realm.objects(MyEvent.self)
        
        // RealmのNotificationを設定
        notificationToken = events.observe { [weak self] (changes: RealmCollectionChange) in
            guard let self = self else { return }
            switch changes {
            case .initial:
                // 初期データの読み込み
                self.calendar.reloadData()
            case .update(_, _, _, _):
                // データの更新
                self.calendar.reloadData()
            case .error(let error):
                // エラー
fatalError("(error)")
}
}

deinit {
    // NotificationTokenを解除
    notificationToken?.invalidate()
}

oggysy replied:
いただいたサンプルコードを元にRealmで用意されている、observeメソッドで変更を監視し画面更新するようにしました。仕組みやこのコードの記法についてまだまだ理解が不十分ですが。。
データが更新された時だけ、カレンダーがリロードされるようになりました！

一点質問です。
現在、今日の日付を赤枠で囲うようにしてます。日付が変ってからカレンダー画面に遷移しても、日付が前の日のままでした。現在の設定だと、データの更新があった時とアプリを起動した時のいずれかしか画面が更新されないと思うのですが、日付が変ったことを通知してカレンダー画面を更新するといった方法はありますでしょうか。調べているのですが、なかなかいい方法が見つかりません。

oggysy committed:
elapsedTimeConvertForDisplay関数の名前変更
setValue関数をupdate(by :)に変更
Conditional Returns on Newline Violationの警告を修正
washDay変数とdisplayIllustration変数を削除し、関数に置き換え
static関数を作ってModel関連のロジックを集約
強制アンラップしている箇所を変更
observeメソッドを使って変更を監視

oggysy committed:
コメントを修正

inushima commented on unoji/Controller/CalendarViewController.swift:
```diff
@@ -0,0 +1,88 @@
+
+    // 日付のセルに画像を表示させる
+    func calendar(_ calendar: FSCalendar, imageFor date: Date) -> UIImage? {
+        df.dateFormat = "yyyy-MM-dd"
```
```suggestion
        let matchObject = objects?.first(where: { $0.formatWashTime() == df.string(from: date) })
        guard let matchObject = matchObject else {
            return nil
        }
        return UIImage(named: matchObject.setDisplayIlustration() + "icon")
```

inushima commented on unoji/Controller/CalendarViewController.swift:
```diff
@@ -0,0 +1,88 @@
+            switch changes {
+            case .initial:
+                // 初期データの読み込み
+                print("initial")
```
print文は消します。

oggysy committed:
日付が変わった時カレンダーをリロードする処理を追加
print文を削除
dataformatの処理をconfigureに移動
calendarのimageFor関数の処理を修正

oggysy committed:
License Plistの追加

inushima reviewed (approved):
細かい点のみコメントしましたが、後はOKです！
お疲れ様でした！

oggysy committed:
calendarのimageFor関数を修正
コメントを修正
